### PR TITLE
[18.01] Build production client

### PR DESCRIPTION
**Pull request by martenson**
with maps

### CODE REVIEW

**Comment by dannon** [comment]
Just occurred to me that we were supposed to discuss sourcemap references at the release meeting, per our last related issue on this.  I'm still opposed to building and shipping files with sourcemap references that *will not work* unless people take extra steps.  It's pointless and wrong, and anyone that's actually going to build and ship sourcemaps will by necessity regenerate these files with references intact.

As of today, dev automatically builds unless you opt out and the expectation is that '`static` will look different' post-checkout-and-run.  We should leverage that expectation, expect that people who want to build and serve sourcemaps are going to understand the build process at least at a superficial level (they must, after all, to build them), and leave these pointless references out of the repo.

**Comment by dannon** [comment]
(adding the maps is a step backward, and not what I meant)

**Comment by martenson** [comment]
I added the maps to complete the PR, not as a reaction to your opposition. 

I still think we should ship the maps with the releases. To me it has hardly any drawbacks comparable to the potential benefits of being able to debug frontend seamlessly.

We should discuss this and I am ready to give up the maps if the consensus is the other way. If this happens I will still try to defend at least official deprecation of this in 18.01 and complete removal in 18.05 though.

**Comment by jxtx** [comment]
> sourcemap references that will not work

@dannon can you explain why the sourcemaps will not work? I'm definitely not in favor of shipping anything that is broken. 

However for release builds, it makes sense to me to include the maps to help people report errors. I'm assuming these are not for developers, but to make it easier do users of production instances running release builds to report any JS errors they see.

**Comment by dannon** [comment]
@jxtx  The plan, as we discussed it last time, was to require deployers to generate the maps on their end as a part of the deployment process, if they wanted to use them.  My point regarding them not working was about the first commit here -- that is, shipping static files with sourcemap references a-la `+//# sourceMappingURL=../../maps/...`, *without* actually shipping those maps.  This mapping information would reference files that weren't there (given that sourcemaps themselves were not included in the repo), and so I didn't think we should ship files with those references -- especially since to generate the maps you'd have to run `make client` or `make client-production-maps`, which would of course regenerate the source files themselves, as well, *with correct mapping references*, along with the maps.

The whole point was to try to get the sourcemaps out of the repo, since they were a relatively large space suck, and are easily generated externally.  The argument has shifted a bit with the addition of the initially forgotten second commit above, it seems, and now it seems like we're wanting to include sourcemaps back in the repo and ship them.  If the plan to have completely stripped sourcemaps from the repo is no longer a given, then I'm fine with adding it all back, but this was not the argument that we were having before, or what my statement above was regarding.

TLDR; if we all want to add the maps back, and ship them, that's fine with me -- just not what I thought we (as a team) agreed the plan was before, nor what my first (pre-second-commit) statement was about.

**Comment by martenson** [comment]
I propose this: for 18.01 we merge this and deprecate shipping maps in the release notes.
This will make it easy for us to drop them in 18.05, when deployers will hopefully be more familiar with client building.

In 18.05 we could also think about hosting these maps for everybody somewhere like `https://depot.galaxyproject.org/<hash>/analysis.js.map`

Also hosting them locally [on sentry](https://docs.sentry.io/clients/javascript/sourcemaps/#uploading-source-maps-to-sentry) might be interesting.

ping @dannon @jmchilton @jxtx

**Comment by dannon** [comment]
@martenson Pushing back to dropping them from releases 'officially' in 18.05 works for me.

How about we also hold off merging this until the official 'release' so we minimize the number of times we're going to rebuild?

**Comment by martenson** [comment]
@dannon yep, this should totally be merged at least after webhooks PR

**Comment by nsoranzo** [comment]
> How about we also hold off merging this until the official 'release' so we minimize the number of times we're going to rebuild?

Doesn't this (using outdated artifacts) affect the CI tests on the release_18.01 branch?

**Comment by martenson** [comment]
@nsoranzo tests should build the client themselves, if they don't we should fix it imo

**Comment by dannon** [comment]
@martenson @nsoranzo Correct, tests always rebuild the client: https://github.com/galaxyproject/galaxy/blob/dev/.ci/jenkins/selenium/run_tests.sh#L20

And the karma-driven unit tests build their own bundles as a part of setup.

**Comment by martenson** [comment]
rebuilt, rebased, ready

**Comment by martenson** [comment]
rebased, rebuilt, ready again

**Comment by natefoo** [comment]
I need this to update Main, so I'm going to go ahead and merge it, we can open additional rebuilds as needed.

**Comment by natefoo** [comment]
On second thought, I just added client build functionality for Test, I guess that during the prerelease period I could just have Main build the client.

**Comment by martenson** [comment]
@natefoo that is a good idea!
